### PR TITLE
Integrate use of `perfplot` package with benchmark

### PR DESCRIPTION
**Pull request by jimthompson5802**
Addresses Issue #193.

@ashvardanian   Here is the start of the PR.  I'm marking it in DRAFT mode because more work is needed on it.

I created a new module `benchmark_perfplot.py` as proof-of-concept of using `perfplot`.  Right now much of using `perfplot` is hard-coded.  For this proof-of-concept, the benchmark compares `np.dot` against `simd.dot` with one dtype.  The plot shows timings for different values of `ndim`.  

As I understand the structure for the original `benchmark.py`, I'm keeping the same structure.  The timing is on how long it takes to process a batch of row vectors, i.e., if the batch is of size `n` and and each row is size `ndim`, then the time reported is the total time it took to process all of the `n` rows of the batch.

Here is a `perfplot` generated for this proof-of-concept.
image

Let me know what you think.

### CODE REVIEW

**Comment by jimthompson5802** [comment]
The `perfplot` specific code in the proof-of-concept, is located here:
https://github.com/jimthompson5802/SimSIMD/blob/1d46ea782f0ae52041b6cbfcacd3cc1d59b63a6d/python/bench_perfplot.py#L312-L345

**Comment by ashvardanian** [comment]
I think it may be a good idea to show relative speedups instead of row duration.

Alternatively, it may be a good idea to show performance stars for different numeric types using NumPy/SciPy performance for double-precision as a baseline.

What do you think?

**Comment by ashvardanian** [comment]
@jimthompson5802 equality comparison logic in such kernels is very tricky and already handled by the test using fixtures. Let’s avoid overcomplicating the benchmark with equality checks.

**Comment by jimthompson5802** [comment]
By speed up, I assume you mean speed up compared to the "baseline" function.  In the proof-of-concept situation, the baseline function is `np.dot` and we should show the speed up for `sims.dot` for the different values of `ndim`.

Is this correct understanding?

**Comment by ashvardanian** [comment]
> By speed up, I assume you mean speed up compared to the "baseline" function. In the proof-of-concept situation, the baseline function is `np.dot` and we should show the speed up for `simd.dot` for the different values of `ndim`.
> 
> Is this correct understanding?

Updated my comment above for clarity 🤗

**Comment by ashvardanian** [comment]
> For this plot, `np.dot` is designated as the "baseline" function, so it speed-up value is "1.0". All the others are relative to `np.dot`. In this situation, `simd.dot` is "faster" because its speed-up factor > 1 for most of the values of `ndim`. OTOH, `spd.cosine` is "slower" because speed-up factor < 1.

We shouldn't put `spd.cosine` and `simd.dot` on the same chart. There should be a separate chart for every function name. In this case if the dot products are reviewed - the `np.dot<float64>` should be the baseline,  but the chart can also feature `np.dot<float16>`, `np.dot<float32>`, `np.dot<int8>`, `simd.dot<float64>`, `simd.dot<float32>`, `simd.dot<float16>`, `simd.dot<bfloat16>`:

> good idea to show performance (speedup) for different numeric types using NumPy/SciPy performance for double-precision as a baseline. So we have one chart for every kernel (be it dot products or cosine distances) with lines for every backend and every numeric type.

We may also want to use the vector size in bytes on the X axis, as opposed to dimensions. Can make a boolean option for that in the CLI 🤗

**Comment by jimthompson5802** [comment]
@ashvardanian   Here is an updated sample plot.  For testing purposes I'm only using on 2 data types per function call.  I believe this is more in-line with what you are looking for.
image

I still have to do more due diligence to confirm the speed-up computations are what you are looking for.  

Let me know what you think.

**Comment by jimthompson5802** [comment]
@ashvardanian I've incorporated what we discussed.  I'm marking this ready for review.  Here are summary of key changes:

* Supporting profiling primarily the `dot` operation for numpy, simsimd, tensorflow, torch and jax.
* numpy.dot<float64> is set as the baseline operation for speed-up comparison
* Bencmark results are returned in a png file called `simsimd_speed_up_numpy.dot(f64).png`

Supported command-line args:
```text
usage: bench_perfplot.py [-h] [--ndim NDIM] [--torch] [--tf] [--jax]

Benchmark SimSIMD

options:
  -h, --help   show this help message and exit
  --ndim NDIM  Size of vectors to benchmark, either 'default' powers of 2 (from 1 to 32K) or comma-seperated list of integers
  --torch      Profile PyTorch
  --tf         Profile TensorFlow
  --jax        Profile JAX
```
 
Here is output for default execution:
```text
# Benchmarking SimSIMD

- Vector dimensions: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768]
- Hardware capabilities: serial, haswell
- SimSIMD version: 5.5.0
- NumPy version: 1.26.4
-- NumPy BLAS dependency: openblas64
-- NumPy LAPACK dependency: dep140213194937296
```
image

Here is output with command line options: `--torch --tf --jax`:
```text
# Benchmarking SimSIMD

- Vector dimensions: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768]
- Hardware capabilities: serial, haswell
- SimSIMD version: 5.5.0
- NumPy version: 1.26.4
- PyTorch version: 2.4.1+cu121
- TensorFlow version: 2.17.0
- JAX version: 0.4.34
-- NumPy BLAS dependency: openblas64
-- NumPy LAPACK dependency: dep140213194937296
```
image

Some items for consideration:
* as I noted in our discussion, `perfplot` does not seem to allow the user to control plot appearance, i.e., unable to change type of line or color or title of plot.  If this level of control is needed, we have to work outside of `perfplot` to display the results.  We can still use `perfplot` to derive the timings.  I've looked at `perfplot`'s internal data structures and can easily extract the timings.  From the extracted timings, we can generate the plots in our own code.
* Do you want a command line option to specify the name of the generated plot image file?

**Comment by jimthompson5802** [comment]
@ashvardanian   I just realized I still have some test/debug code that affects timing of the function that is being tested.  Changed the PR back to DRAFT.  I should have that code cleaned up by tonight.

**Comment by jimthompson5802** [comment]
@ashvardanian   Finish  code clean-up.  Ready for your review.

With this clean, reduced overhead for collecting timings.  Here is the updated default execution:
```text
# Benchmarking SimSIMD

- Vector dimensions: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768]
- Plot file path: simsimd_speed_up.png
- Hardware capabilities: serial, haswell
- SimSIMD version: 5.5.0
- NumPy version: 1.26.4
-- NumPy BLAS dependency: openblas64
-- NumPy LAPACK dependency: dep140213194937296
```
image

You should notice a significant change in the reported speed-up factors for the lower range of `ndim` values.

I added two additional cli options:
* `plot_fp`: specifies the output plot file name.  Default is `simsimd_speed_up.png`
* `debug`:  provides useful information when testing new kernels.  This will effect speed-up factors, especially on the smaller range of `ndim`, as noted earlier.

Here are all the cli options:
```text
usage: bench_perfplot.py [-h] [--ndim NDIM] [--torch] [--tf] [--jax] [--plot_fp PLOT_FP] [--debug]

Benchmark SimSIMD

options:
  -h, --help         show this help message and exit
  --ndim NDIM        Size of vectors to benchmark, either 'default' powers of 2 (from 1 to 32K) or comma-seperated list of
                     integers
  --torch            Profile PyTorch
  --tf               Profile TensorFlow
  --jax              Profile JAX
  --plot_fp PLOT_FP  File to save the plot to, default: 'simsimd_speed_up.png'
  --debug            Provide additional debug information
```
